textsearch/python/textsearch/suffix_array.py

```diff
@@ -0,0 +1,107 @@
+# Copyright      2023   Xiaomi Corp.       (author: Wei Kang)
+#
+# See ../../../LICENSE for clarification regarding multiple authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import _fasttextsearch
+import numpy as np
+
+
+def create_suffix_array(input: np.ndarray) -> np.ndarray:
+    """
+    Creates a suffix array from the input text and returns it as a NumPy array.  Read
+    the usage carefully as it has some special requirements that will require careful data
+    preparation.
+
+    Args:
+       input: an integer (or unsigned integer) type of np.ndarray.  Its shape
+          should be (seq_len + 3,) where `seq_len` is the text sequence length INCLUDING
+          EOS SYMBOL.
+          The EOS (end of sequence) symbol must be the second largest element of the
+          type (i.e. of the form 2^n - 2), must be located at input[seq_len - 1] and
+          must appear nowhere else in `input` (you may have to map the input
+          symbols somehow to achieve this).  It must be followed by 3 zeros, for reasons
+          related to how the algorithm works.
+    Returns:
+          Returns a suffix array of type np.int64,
+          of shape (seq_len,).  This will consist of some permutation of the elements
+          0 .. seq_len - 1.
+    """
+    assert input.ndim == 1, input.ndim
+    seq_len = input.size - 3
+    assert seq_len >= 1, seq_len
+    max_symbol = input[seq_len - 1]
+    assert max_symbol == np.iinfo(input.dtype).max - 1, max_symbol
+    assert np.alltrue(input[seq_len:] == np.array([0, 0, 0], dtype=input.dtype))
+
+    # The C++ code requires the input array to be contiguous.
+    input64 = np.ascontiguousarray(input, dtype=np.int64)
```

Comment:
Is there a reason to use `int64` here?

All Unicode code points can be stored in 3 bytes.
`int32` is sufficient in this task for the input array, I think.

Reply:
The elements in suffix_array are the indexes, I think.

Reply:
Dan suggested not to change the code in C++ as it already tested, in original code, only one template (i.e. elements and indexes share the template).

Reply:
I am referring to the input argument, i.e., the element type in the input array, not the element in the return value.

Reply:
> I am referring to the input argument, i.e., the element type in the input array, not the element in the return value.

Yes, I know, I said they share the same template argument in original code, pls read all the comments above.